annonatate/utils/github.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import os, json
+import base64
+from flask_github import GitHub
+
+import pdb
+
+class GitHubAnno(GitHub):
+  def setbork(self):
+    self.bork = 'bork'
+
+  def get_existing(self, session, full_url):
+      payload = {'ref': session['github_branch']}
+      match = False
+      if '/images/' in full_url:
+          full_url, match = full_url.strip('/').rsplit('/', 1)
+      existing = self.raw_request('get',full_url, params=payload).json()
+      if match and type(existing) == list:
+          matches = list(filter(lambda x: x['name'] == match, existing))
+          existing = matches[0] if len(matches) > 0 else matches
+      if 'sha' in existing:
+          return existing['sha']
+      else:
+          return ''
+
+  def sendgithubrequest(self, session, filename, annotation, path, order=''):
+      data = self.createdatadict(session, filename, annotation, path, order)
+      response = self.raw_request('put', data['url'], data=json.dumps(data['data'], indent=4))
+      return response
+
+  def createdatadict(self, session, filename, text, path, order=''):
+      full_url = os.path.join(session['github_url'], path, filename)
+      sha = self.get_existing(session, full_url)
+      writeordelete = "write" if text != 'delete' else "delete"
+      message = "{} {}".format(writeordelete, filename)
+      text = '---\ncanvas: "{}"\n{}---\n{}'.format(text['target']['source'],order, json.dumps(text, indent=4)) if type(text) != str and type(text) != bytes else text
+      text = text.encode('utf-8') if type(text) != bytes else text
+      data = {"message":message, "content": base64.b64encode(text).decode('utf-8'), "branch": session['github_branch'] }
+      if sha != '':
+          data['sha'] = sha
+      return {'data':data, 'url':full_url}
+
+  def updateAnnos(self, session):
```

Comment:
moving this out is making it not useful in terms of functionality. The main reason it is there is to update the annotations faster than the api. It is updating the session['annotations'], something it can not do in this utils file.

Reply:
Hmm, I didn't see that. I'll have another look. It could move back into the main app.

Reply:
I think this should be ok: updateAnnos updates the ```session``` that is passed to it, which would normally be the main session (e.g. in flaskserver.py at ```githubresponse = github.updateAnnos(session)```. I'm just adding a test to confirm this.

Reply:
updateAnnos is used twice. It causes problems where it does not have an equal.

Reply:
I am thinking I might want to get rid of this. I am not sure how useful it is. It was put in place to solve two problems. If two people are working on the application at the same time during the period where we are not querying the API, it would grab any annotations from the GitHub repo so both could see what the other is doing. 
The other is to remove any deleted annotations that did not get deleted in API because it is still rebuilding. I would love to hear your thoughts.

Reply:
I'd vote for keeping it - it doesn't look fragile, and while it might not come into play very often, when it does it might protect someone from having their data messed up. I haven't thought about live collaborative annotating, but I can imagine it would be a cool way for work for some projects.